src/contract.rs

```diff
@@ -242,10 +248,103 @@ pub fn query<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>, msg: QueryM
         QueryMsg::ContractStatus {} => query_contract_status(&deps.storage),
         QueryMsg::ExchangeRate {} => query_exchange_rate(&deps.storage),
         QueryMsg::Minters { .. } => query_minters(deps),
+        QueryMsg::BalanceWithPermit { signed, signature } => {
+            query_balance_with_permit(deps, signed, signature)
+        }
         _ => authenticated_queries(deps, msg),
     }
 }
 
+fn query_balance_with_permit<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    signed: SignedPermit,
+    signature: Signature,
+) -> Result<Binary, StdError> {
+    if signed.msgs.len() != 1 {
+        return Err(StdError::generic_err(format!(
+            "Must sign exactly 1 permit message, got: {:?}.",
+            signed.msgs.len()
+        )));
+    }
+
+    if signed.msgs[0].r#type != "query_balance_permit" {
+        return Err(StdError::generic_err(format!(
+            "Type must be 'query_balance_permit', got: {:?}.",
+            signed.msgs[0].r#type
+        )));
+    }
+
+    let permit = &signed.msgs[0].value;
+
+    // Validate permit_user_id
+    let _permit_user_id = &permit.permit_user_id;
+    // TODO fail if permit_user_id is revoked
+
+    // Validate permit message
+    let query_balance_permit_msg = ReadonlyConfig::from_storage(&deps.storage)
+        .constants()?
+        .query_balance_permit_msg;
+    if permit.message != query_balance_permit_msg {
+        return Err(StdError::generic_err(format!(
+            "Message must be '{:?}', got: '{:?}'.",
+            query_balance_permit_msg, permit.message
+        )));
+    }
+
+    // Validate that account is derived from pubkey
+    let pubkey = signature.pub_key.value;
+    let account_from_pubkey = deps.api.human_address(&pubkey_to_account(&pubkey))?;
+
+    if account_from_pubkey != permit.query_balance_of {
+        return Err(StdError::generic_err(format!(
+            "Input account {:?} not the same as account {:?} that was derived from pubkey {:?}.",
+            permit.query_balance_of,
+            account_from_pubkey,
+            pubkey.clone()
+        )));
+    }
```

Comment:
should we use a constant time comparison here?

Reply:
No